feat: Implement Enhanced Syslog for Dagger-Registry Redirect and Add Local Testing Guide

grouville opened the pull request:
Fixes https://linear.app/dagger/issue/INFRA-36/mitigate-registrydaggerio-nats-spof

## Problem
@gerhard and @jlongtine found out that when Fly.io NATS is affected, our ability to report metrics becomes affected too. See [the original incident](https://uptime.betterstack.com/team/132610/incidents/415909666f) for more details.

## Solution
This PR extends the app's logger to write to syslog as well as the current stdout.
As these are important changes, a markdown file `Local Testing Guide for Dagger-Registry Redirect.md` explains how to test the app.

## Pre-requisites
- [x] Merge https://github.com/dagger/dagger.io/pull/2746
- [x] Merge https://github.com/dagger/dagger.io/pull/2750

## Follow-up
- [x] update Magefile to deploy this app in parallel in `fly.io` (taking into consideration the new API v2 changes)
- [x] update corresponding Vector files: https://github.com/dagger/dagger.io/pull/2735, verify filter's app name (review listening interface)
- [x] one last manual check that everything behaves correctly
- [ ] point registry.dagger.io to the new app instance
- [ ] delete previous registry-redirect instance in `~24h` when we have the confidence that everything behaves as expected

grouville committed:
feat: Implement Enhanced Syslog for Dagger-Registry Redirect and Add Local Testing Guide

This commit introduces several enhancements and new features:

- Implemented a new Syslog in pkg/logger/logger.go with accompanying unit tests in logger_test.go.
- Introduced the Syslog writer in pkg/syslogger/syslogger.go, along with a mock_syslogger for testing purposes. Unit tests for the syslogger are available in syslogger_test.go.
- Updated main.go to integrate the new Syslog functionalities.
- Made relevant updates to dependencies in go.mod and go.sum.
- Added a comprehensive local testing guide (Local Testing Guide for Dagger-Registry Redirect.md) to facilitate local testing and verification of the Dagger-Registry Redirect.

These improvements enhance the logging capabilities of the Dagger-Registry Redirect and streamline the process of local testing.

Signed-off-by: grouville <guillaume@dagger.io>

gerhard commented on main.go:
```diff
@@ -49,11 +51,23 @@ func main() {
+		Level: "info",
+	}
+
+	ctx, syslogger, err := logger.SetupLogging(ctx, logCfg, "dagger-registry-2023-01-23")
```
`dagger-registry-2023-01-23` should be an environment variable. Maybe we can use `FLY_APP_NAME`?

gerhard commented on pkg/logger/logger.go:
```diff
@@ -0,0 +1,73 @@
+	"knative.dev/pkg/logging"
+)
+
+type MyLoggingConfig struct {
```
Since this is in the `logger` package, `Config` seems sufficient. `logger.Config` is short & expressive.

gerhard commented on pkg/syslogger/syslogger.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+func NewSyslogWriter(priority syslog.Priority, tag string) (*SyslogWriter, error) {
+	w, err := syslog.Dial("udp", "0.0.0.0:514", priority, tag)
```
We want to use `TCP` so that we don't lose log messages (UDP is fire & forget). If the logger cannot establish a TCP connection, it should panic - [let it crash](https://wiki.c2.com/?LetItCrash). We wan to be sure that the other end is able to receive these logs.

We also want the `HOST:PORT` to come from an environment variable.

The whole purpose of this PR is so that we don't lose log messages. Investing in integration - good thing we have Dagger! - is worth the effort 💪

grouville committed:
chore: simplify implementation

- Remove interface, as no mockups are required
- Rename and simplify code
- Modify syslogger unit tests (race conditions)
- Add logger integration test (currently not testing http server graceful shutdown)

Signed-off-by: grouville <guillaume@dagger.io>

grouville committed:
fix: E2E logger_test

Still a race on the logger, I wonder what the best synchronization strategy is the best option:
- a wg group at the handler level
- a buffer at the logger level

Signed-off-by: grouville <guillaume@dagger.io>

grouville committed:
chore: test app on fly

Test deploying the app on fly, redirect to ngrok -> local vector instance

Signed-off-by: grouville <guillaume@dagger.io>

grouville committed:
fix: full E2E test of our shutdown process

E2E integration test in which we test our shutdown logic to see if we lose data when having a concurrent shutdown signal and incoming streams. In this test, we setup a local HTTP and TCP server (fake syslogger), send X messages to the HTTP server and right after that send the HTTP shutdown signal (faked with a channel).

The HTTP server mimicks our main.go implementation: it wraps the handler with a waitGroup and waits for all processed logs to be written in the HTTP shutdown goroutine.

Signed-off-by: grouville <guillaume@dagger.io>

grouville committed:
chore: update docs + retrieve app_name from env

- Update docs to follow new testing methodology
- Implement app name retrieval from env

Signed-off-by: grouville <guillaume@dagger.io>

gerhard commented on magefiles/dagger.go:
```diff
@@ -189,7 +188,7 @@ func deploy(ctx context.Context, c *dagger.Client, imageRef string) {
 // [lints, tests], builds, publishes & deploys a new version of the app
 func All(ctx context.Context) {
-	mg.CtxDeps(ctx, Lint, Test)
+	// mg.CtxDeps(ctx, Lint, Test)
```
uncomment before merging

gerhard commented on Local Testing Guide for Dagger-Registry Redirect.md:
````diff
@@ -0,0 +1,86 @@
+2. Run the following command to setup a reverse proxy:
+
+```shell
+caddy reverse-proxy --from toto.localhost --to http://localhost:8080
+```
+
+### Running the Main Application
+
+With the environment set up, run the Dagger-Registry Redirect application:
+
+```shell
+go run . -repo dagger
+```
+
+You should see the logs locally and in the Vector instance.
+
+To check the next logs, run the following Docker command (assuming the [equinix](https://github.com/orgs/dagger/packages/container/package/equinix-demo-day-2023) package exists). This command does not interfere with production as the registry-redirect is running locally and is separate from our Vector configurations.
+
+```shell
+docker pull toto.localhost/equinix-demo-day-2023:latest
+```
+
+## Fly Procedure
+
+### Deploy the app to FLY
+```shell
+cd magefiles && FLY_APP_NAME="dagger-registry-2023-07-28" FLY_API_TOKEN="$(flyctl auth token)" GITHUB_REF_NAME=main  go run ./main.go -w .. all
+```
+
+### Trigger logs
+
+To trigger the logs, run the following Docker command (assuming the [equinix](https://github.com/orgs/dagger/packages/container/package/equinix-demo-day-2023) package exists).
+
+1. Find the app name's hostname: https://fly.io/apps/dagger-registry-2023-07-28 in our case.
+
+2. Run below command:
+
+```shell
+docker pull dagger-registry-2023-07-28.fly.dev/equinix-demo-day-2023:latest
+```
````
We should either capture this in dagger, or simply remove. It seems a lengthy process with various dependencies that may be reasonable to set up for development purposes. For others to test this locally, a mage target (a.k.a. daggerize) sounds much lower friction.

grouville replied:
Captured it here https://gist.github.com/grouville/e772c61a4213d51eac8cb9cabdea11f3

gerhard commented on magefiles/dagger.go:
```diff
@@ -26,7 +25,7 @@ const (
 	flyctlVersion = "0.1.65"
 
-	appName          = "dagger-registry-2023-01-23"
+	appName          = "dagger-registry-2023-07-28"
```
this will likely need a few other changes in our vector config, we can go through them together when we are ready to merge.

gerhard commented on main.go:
```diff
@@ -49,6 +52,20 @@ func main() {
+		Level:     "info",
+		Component: os.Getenv("FLY_APP_NAME"),
+		Protocol:  "tcp",
+		Address:   "7.tcp.eu.ngrok.io:18985", // 0.tcp.eu.ngrok.io:16901
```
read from env, same as app name above. `SYSLOG_HOST` reads well to me.

gerhard commented on main.go:
```diff
@@ -49,6 +52,20 @@ func main() {
+	// info, warning, error and fatal
+	logCfg := logger.Config{
+		Level:     "info",
+		Component: os.Getenv("FLY_APP_NAME"),
```
if this needs to be set, consider using `panic("FLY_APP_NAME env var must be set")` if the value is `""`

gerhard commented on pkg/logger/logger_test.go:
```diff
@@ -0,0 +1,246 @@
+func makeLoggingHandler(ctx context.Context) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		logger := logging.FromContext(ctx)
+		select {
+		case <-time.After(5 * time.Second):
+			// simulate a long running request, in case ctx.Done() (srv.ShutdownServer)
+			// is not called fast enough
+		case <-ctx.Done():
+			break
+			// the context will be canceled when the server is shutting down
+			// this gives an exact timing of workload simulation
+		}
```
this will exit the select after at most 5 seconds, but also when `ctx.Done()` is triggered. I propose that we make `ctx.Done()` the first statement (no need to `break`) & the `time.After()` second. Timing out these requests after `1s` seems a better lower value (`5s` is long time to wait in tests).

gerhard commented on pkg/logger/logger_test.go:
```diff
@@ -0,0 +1,246 @@
+	}()
+
+	// Wait until the server is ready
+	timeout := time.After(5 * time.Second)           // Wait up to 5 seconds for server to start
```
http server should be ready within `1s`. something is wrong if it takes longer than this.

gerhard commented on pkg/logger/logger_test.go:
```diff
@@ -0,0 +1,246 @@
+
+	<-httpReady // Wait until the http server is ready
+
+	// Send 100 requests to the http server
```
remove all hard-coded values from comments. if they change (happens more often than you think) the comments will be outdated.

gerhard commented on pkg/logger/logger_test.go:
```diff
@@ -0,0 +1,246 @@
+	close(shutdownHTTP)
+
+	// Count received messages
+	count := 0
+	for range messages {
+		count++
+	}
```
can you run a `len` on the channel instead of looping through all the messages? in that case, you may not even need a variable, use it directly in the comparison.

gerhard reviewed (commented):
Looking good, almost there!

grouville committed:
feat: implement Gerhard's fixes following review

Update handler's logic, cleanup test, make app more generic with env variables

Signed-off-by: grouville <guillaume@dagger.io>
Co-authored-by: Gerhard Lazu <gerhard@dagger.io>

grouville committed:
fix: race when not waiting for all messages to have been processed

When closing the syslog server, I was not signaling the main goroutine that all messages had been processed.
Also, easier implementation of httpserver

Signed-off-by: grouville <guillaume@dagger.io>

grouville committed:
fix: update logger formatting

Rely on JSON formatting for easier processing on Vector's side

Signed-off-by: grouville <guillaume@dagger.io>

grouville committed:
feat: implement reconnection mechanism

Implements the reconnection mechanism in logger's logic, when losing the TCP connection

Signed-off-by: grouville <guillaume@dagger.io>

grouville committed:
chore: update deployment

Signed-off-by: grouville <guillaume@dagger.io>

grouville committed:
fix: segfault on syslog writer when not connected

Guard against a segfault when the writer hasn't connected

Signed-off-by: grouville <guillaume@dagger.io>

Commit:
Use run time env vars, not build time

Also sensible default values when running locally, nothing
platform-specific.

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

Commit:
Capture syslogHost in our Dagger pipeline

So that this is set at deploy time & we have a single source of truth.

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

Commit:
Add pipeline labels & update all deps to latest (except Go)

Only deploy when the app source has changed (including the pipeline code).

Remove logs target since it did not prove useful to have - we run this
command locally when we need it.

Share the same client when running multiple targets via All

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

Commit:
Cache go deps from magefiles

We do not use the top-level go deps for GHA, everything runs in Dagger.

Signed-off-by: Gerhard Lazu <gerhard@dagger.io>

gerhard reviewed (approved):
👍 🚀
